File: packages/pip/src/goodvibes_cli/steps/project_copies.py
"""Goodvibes copies a project keeps from project scope once goodvibes is set up for all projects."""
from __future__ import annotations

import hashlib
import pathlib

from goodvibes_cli.steps.write_manifest import USER_OWNED, USER_REMOVED
from goodvibes_cli.utils.safe_path import SymlinkError, check_writable

STRIPPED = "CLAUDE.md: removed the old goodvibes rules block; the rules now come from your Claude Code settings folder"
STRIP_PLAN = "CLAUDE.md: will remove the old goodvibes rules block; the rules now come from your Claude Code settings folder"
EDITED = "Edited skill copies stay in this project; the same skills are now set up for all your projects, so Claude may load both. Delete a copy you no longer need: "


def removed_line(rel: str) -> str:
    return f"{rel}: removed, now set up for all your projects"


def old_skill_copies(cwd: pathlib.Path, files: dict[str, str]) -> tuple[list[str], list[str]]:
    """Tracked project skill files: (unchanged since goodvibes wrote them, edited by the user)."""
    unedited: list[str] = []
    edited: list[str] = []
    for rel, sha in files.items():
        if not rel.startswith(".claude/skills/") or sha in (USER_OWNED, USER_REMOVED):
            continue
        try:
            check_writable(cwd, cwd / rel)
        except SymlinkError:
            continue
        if (cwd / rel).is_file():
            (unedited if hashlib.sha256((cwd / rel).read_bytes()).hexdigest() == sha else edited).append(rel)
    return sorted(unedited), sorted(edited)

File: packages/npm/src/steps/project-copies.ts
import { readFile, stat } from 'node:fs/promises'
import { join } from 'node:path'
import { createHash } from 'node:crypto'
import { USER_OWNED, USER_REMOVED } from './write-manifest.js'
import { writeBlocked } from '../utils/fs-safe.js'

export const STRIPPED = 'CLAUDE.md: removed the old goodvibes rules block; the rules now come from your Claude Code settings folder'
export const STRIP_PLAN = 'CLAUDE.md: will remove the old goodvibes rules block; the rules now come from your Claude Code settings folder'
export const EDITED = 'Edited skill copies stay in this project; the same skills are now set up for all your projects, so Claude may load both. Delete a copy you no longer need: '
export const removedLine = (rel: string) => `${rel}: removed, now set up for all your projects`

// Tracked project skill files left from project scope: unchanged since goodvibes wrote them, or edited by the user.
export async function oldSkillCopies(cwd: string, files: Record<string, string>): Promise<{ unedited: string[]; edited: string[] }> {
  const unedited: string[] = []
  const edited: string[] = []
  for (const [rel, sha] of Object.entries(files)) {
    if (!rel.startsWith('.claude/skills/') || sha === USER_OWNED || sha === USER_REMOVED) continue
    if (await writeBlocked(cwd, rel)) continue
    const path = join(cwd, rel)
    if (!(await stat(path).then(s => s.isFile(), () => false))) continue
    const got = createHash('sha256').update(await readFile(path)).digest('hex')
    ;(got === sha ? unedited : edited).push(rel)
  }
  return { unedited: unedited.sort(), edited: edited.sort() }
}
